hpe3parclient/client.py

```diff
@@ -4816,3 +4731,159 @@ def getScheduleStatus(self, schedule_name):
+            ret = {
+                'total': 5,
+                'members': [
+                    {
+                        'Package': 'Complete',
+                        'Version': '3.1.2.422',
+                        'InstallTime': '2013-08-21 18:06:45 PDT',
+                        'Id': 'MU2'
+                    },
+                    {
+                        'Package': 'Complete',
+                        'Version': '3.1.2.484',
+                        'InstallTime': '2013-10-10 15:20:05 PDT',
+                        'Id': 'MU3'
+                    },
+                    {
+                        'Package': 'Complete',
+                        'Version': '3.1.3.170',
+                        'InstallTime': '2014-01-30 11:34:20 PST',
+                        'Id': 'DEVEL'
+                    },
+                    {
+                        'Package': 'Complete',
+                        'Version': '3.1.3.202',
+                        'InstallTime': '2014-03-26 13:59:42 PDT',
+                        'Id': 'GA'
+                    },
+                    {
+                        'Package': 'Complete',
+                        'Version': '3.1.3.230',
+                        'InstallTime': '2014-06-06 14:46:56 PDT',
+                        'Id': 'MU1'
+                    }
+                ]
+            }
+
+            # Example 2: No data rows, so zero members.
+            ret = {'total': 0, 'members': []}
+
+        """
+
+        members = []
+        if cli_output and len(cli_output) >= 2:
+            for index, line in enumerate(cli_output):
+                if index == 0:
+                    headers = line.split(',')
+                else:
+                    split = line.split(',')
+                    member = {}
+                    for i, header in enumerate(headers):
+                        try:
+                            member[header] = split[i]
+                        except IndexError:
+                            member[header] = None
+                    members.append(member)
+
+        return {'total': len(members), 'members': members}
+
+    @staticmethod
+    def getPortNumber(ip, login, password, port=22,
+                      conn_timeout=None, privatekey=None,
+                      **kwargs):
+        """Get port number from showwsapi output
+
+        :param 3PAR credentials
+        :return: HTTPS_Port column value
+        """
+        ssh_client = ssh.HPE3PARSSHClient(ip, login, password, port,
+                                          conn_timeout, privatekey,
+                                          **kwargs)
+        if ssh_client is None:
+            raise exceptions.SSHException('SSH is not initialized. Initialize'
+                                          ' it by calling "setSSHOptions".')
+        ssh_client.open()
+        cli_output = ssh_client.run(['showwsapi'])
+        if ssh_client:
+            ssh_client.close()
```

Comment:
Preferred way of doing open/close is to put open() call in try block and close() call in finally block just to ensure close() is called even if any exception occurs before reaching this point.

Reply:
> Preferred way of doing open/close is to put open() call in try block and close() call in finally block just to ensure close() is called even if any exception occurs before reaching this point.

Finally is already handled in ansible Module as given below:-
    finally:
        client_obj.logout()

Reply:
login/logout is happening in try/finally. But as I can see ssh_client.open() and ssh_client.close() are outside try/finally. If that's the case, can we use try/finally for open/close as well?